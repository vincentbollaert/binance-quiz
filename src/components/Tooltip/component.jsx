import React, { useState } from 'react'
import { string, bool, number, oneOfType } from 'prop-types'
import styled from 'styled-components'
import { UNIT_LG, UNIT_MD, MEDIUM_AQUAMARINE, FONT_SIZE_SM } from '../../styles'

const Wrap = styled.div`
  position: absolute;
  right: ${UNIT_LG};
  cursor: pointer;
`
const Label = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  border: 1px solid ${MEDIUM_AQUAMARINE};
  width: 24px;
  height: 24px;
  color: ${MEDIUM_AQUAMARINE};
  border-radius: 50%;
  font-size: ${FONT_SIZE_SM};
`
const Tooltip = styled.div`
  visibility: hidden;
  position: absolute;
  right: 100%;
  top: 50%;
  border-right: 2px solid ${MEDIUM_AQUAMARINE};
  margin-right: ${UNIT_MD};
  padding: 1rem ${UNIT_LG};
  line-height: 1.1;
  font-size: ${FONT_SIZE_SM};
  text-transform: uppercase;
  background-color: ${MEDIUM_AQUAMARINE};
  color: rgba(0, 0, 0, 0.6);
  font-weight: bold;
  border-radius: 2px;
  white-space: nowrap;
  transform: translateY(-50%);

  &::before {
    content: "";
    position: absolute;
    top: 50%;
    left: 100%;
    border-style: solid;
    border-width: 6px 0 6px 6px;
    border-color: transparent transparent transparent ${MEDIUM_AQUAMARINE};
    width: 0;
    height: 0;
    transform: translateY(-50%);
  };

  ${props => props.isVisible && `
    visibility: visible;
    margin-right: ${UNIT_LG};
    transition: margin 0.2s ease-out, visibility 0.4s ease-out;
  `};
`

const TooltipWrap = ({ isShow, label, tooltip, className }) => {
  const [isVisible, setIsVisible] = useState(false)

  return (
    !isShow ? null : (
      <Wrap className={className} onClick={() => setIsVisible(!isVisible)}>
        <Label>{label}</Label>
        <Tooltip isVisible={isVisible}>
          {tooltip}
        </Tooltip>
      </Wrap>
    )
  )
}

TooltipWrap.propTypes = {
  isShow: bool.isRequired,
  label: oneOfType([string, number]).isRequired,
  tooltip: string.isRequired,
  className: string,
}

TooltipWrap.defaultProps = {
  className: undefined,
}

export default TooltipWrap
