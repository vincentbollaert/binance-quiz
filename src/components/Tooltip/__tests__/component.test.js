import '@testing-library/jest-dom'
import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import Tooltip from '../component'

describe('Tooltip', () => {
  const props = {
    isShow: false,
    label: 'label',
    tooltip: 'tooltip',
  }
  test('matches snapshot', () => {
    const { container } = render(<Tooltip {...props} />)
    expect(container.firstChild).toMatchSnapshot()
  })
  test('renders nothing by default', () => {
    const { queryByTestId } = render(<Tooltip {...props} />)
    expect(queryByTestId('component-tooltip')).toBeFalsy()
  })
  test('renders on isShow with hidden tooltip', () => {
    const { queryByTestId, queryByText } = render(<Tooltip {...props} isShow />)
    expect(queryByTestId('component-tooltip')).toBeTruthy()
    expect(queryByText('tooltip')).not.toBeVisible()
    expect(queryByText('label')).toBeTruthy()
  })
  test('shows tooltip on toggle click', () => {
    const { container, queryByText } = render(<Tooltip {...props} isShow className="class" />)
    fireEvent.click(container.querySelector('.class'))
    expect(queryByText('tooltip')).toBeVisible()
    fireEvent.click(container.querySelector('.class'))
    expect(queryByText('tooltip')).not.toBeVisible()
  })
})
