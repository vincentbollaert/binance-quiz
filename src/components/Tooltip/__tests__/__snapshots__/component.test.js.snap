// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip matches snapshot 1`] = `null`;
